[1단계 - 지하철 노선도 미션] 파노(전환오) 미션 제출합니다.

jho2301 opened the pull request:
안녕하세요 @JaeYeopHan 님
 체프 @Puterism  와 같이 과제 진행하고 있는 파노입니다!
1단계 과제 기능 구현 완료해서 PR올립니다.
잘 부탁드립니다..!

## 데모 (not working)
https://protected-temple-40587.herokuapp.com/
- Mixed Content 이슈
![image](https://user-images.githubusercontent.com/44419181/111754516-6fcf3d00-88db-11eb-97ed-e79e8cc18a14.png)


## 웹 VSCode 환경
https://github.surf/jho2301/javascript-subway/tree/hwano-cheffe-step1

## 디렉토리 구조
``` plainText
./src/js
├── api
│   └── httpClient.js
├── constants.js
├── eventHandlers
│   ├── handleLogin.js
│   ├── handleLogout.js
│   ├── handleRoute.js
│   └── handleSignup.js
├── index.js
├── routes
│   ├── index.js
│   ├── routeHandler.js
│   └── utils.js
├── services
│   └── auth.js
├── store
│   ├── accessToken.js
│   └── index.js
├── templates
│   ├── lines.js
│   ├── login.js
│   ├── map.js
│   ├── search.js
│   ├── sections.js
│   ├── signup.js
│   └── stations.js
└── utils
    ├── dom.js
    ├── history.js
    ├── localStorage.js
    └── mock.js

7 directories, 24 files
```

## 기능 구현 목록
 ### Webpack을 통한 번들링

- [x] Webpack에서 babel을 설정한다.
- [x] 기타 플러그인 및 설정은 본인이 필요에 따라 추가한다.

### 라우팅 기능

- [x] Browser History Api를 이용하여 SPA처럼 라우팅을 적용한다.

### 회원 기능

- [x] 유저는 회원 가입을 할 수 있다,
  - [x] 회원 가입시 받는 정보는 `email`, `name`, `password`이다.
- [x] 유저는 로그인 할 수 있다.
  - [x] 로그인한 유저만, 서비스에 접근이 가능하다.
  - [x] 로그인하지 않은 유저에게는 로그인 페이지로 리다이렉트 된다.
  - [x] 로그인하면 `/` 루트 페이지로 이동한다.
- [x] 유저는 로그아웃할 수 있다.
  - [x] 로그아웃하면 `/login` 로그인 페이지로 돌아온다.

## 신경 쓴 부분

1. authenticateRouter, unauthenticateRouter 함수를 만들어서 라우터가드 로직을 만들어봤습니다. 
2.  기존 패턴에 대한 논의 없이 과제 요구사항을 수행하면서 필요할 때마다 리팩토링을 통해 구조를 잡아 나갔습니다. 관련해서 팁이 있다면 조언 부탁드리고 싶습니다.
3. httpClient를 추상화해서 분리를 해보았습니다
4. pushState 커스텀이벤트로 프로그램 내 자유로운 라우팅이 가능하도록 의도해봤습니다

## 질문 사항

- addEventListener로 걸어준 이벤트 대상 DOM 객체가 삭제되거나 detach되서 더 이상 참조되지 않으면,  일일이 removeEventListener를 해주는게 좋을까요? 아래 글을 보았는데 애매해서 질문드립니다.

    [https://stackoverflow.com/questions/4337582/do-events-handlers-on-a-dom-node-get-deleted-with-the-node](https://stackoverflow.com/questions/4337582/do-events-handlers-on-a-dom-node-get-deleted-with-the-node)

- SPA를 구현할 때, 다른 페이지의 HTML 코드를, 별도의 html 파일로 만드는 게 좋을까요? 아니면 js 파일에서 template 문자열로 가지고 있게 하는 것이 좋을까요?

![image](https://user-images.githubusercontent.com/44419181/111756498-9c845400-88dd-11eb-9998-226f56841fa9.png)


- 위 코드는 html 파일을 fetch해 와서 문자열로 변환한 뒤, innerHTML로 렌더하는 예시 코드입니다. 이 형식이 갖는 이점과 template 문자열을 js로 관리해서 렌더하는 것과 어떤 차이가 있는지 궁금합니다.

- mini-css-extract-plugin이 어떤 필요에 의해서 만들어졌는지 궁금합니다. css 파일로 분리해서 번들링을 해주는 것은 좋은데, html에 style 태그로 주입되는 것과, css 파일을 link 태그로 가져오는 것과 성능적으로나 기능적으로나 어떤 차이가 생기는지 궁금합니다. 이전에는 html 파일과 css 파일을 따로 분리해서 개발하는 것이 당연하지만, 번들링을 해주니 굳이 css 파일을 분리해도 되는지 잘 모르겠습니다.

jho2301 committed:
chore: project init

jho2301 committed:
chore: prettier 세팅

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
chore: eslint 세팅

jho2301 committed:
chore: cypress 세팅

jho2301 committed:
chore: webpack 세팅

- babel-loader
- dev-server
- css-loader

jho2301 committed:
docs: 라우팅 기능구현목록 구체화

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
chore: gitignore 케이스 추가

- cypress/fixtures/
- cypress/plugins/
- cypress/support/

jho2301 committed:
test: [역 관리] 메뉴를 클릭하면, 역 관리 페이지가 렌더된다.

jho2301 committed:
docs: 기능 구현 목록 추가

- 1단계 UI/UX

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: Browser History Api를 이용하여 SPA처럼 라우팅을 적용한다.

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 회원 가입 기능 구현

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 인증 기능 구현

- 기능 구현 목록 수정
- 인증에 따른 라우터가드 구현

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
refactor: 함수 분리 및 함수명 변경

함수명 변경
- pushState -> routeTo
- checkAuth -> authenticatedRoute

jho2301 committed:
feat: UI / UX 개선

- 입력 폼이 보여질 때, autofocus 되도록 한다.
- 버튼에 hover하거나 focus 되었을 때의 스타일을 적용한다.

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
fix: 로그인 실패 시 alert이 호출되지 않는 현상

jho2301 committed:
fix: 회원가입 후 자동 로그인 안됨

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 committed:
feat: 로그인 상태 유지 선택 기능

jbee37142 commented on webpack.config.js:
```diff
@@ -0,0 +1,42 @@
+        exclude: /node_modules/,
+        use: {
+          loader: 'babel-loader',
+          options: {
+            presets: ['@babel/preset-env'],
+            plugins: ['@babel/plugin-transform-runtime'],
+          },
```
babelrc 파일이 있으면 그 값을 참조하지 않나요?

jho2301 replied:
webpack.config파일에 바벨로더의 옵션이 있으면 rc파일이 없어도 되는 것은 확인했는데 반대의 경우도 동작하네요! 
웹팩 공식문서에서 다시 확인해봤는데 관련 내용을 찾기가 어렵네요ㅜㅜ 웹팩 어려운 것 같습니다. 더 노력하겠습니다.

jbee37142 commented on src/js/routes/utils.js:
```diff
@@ -0,0 +1,22 @@
+import { routeTo } from '../utils/history';
+
+export const authenticatedRoute = async route => {
+  const isLogin = await checkLogin();
```
라우트를 이동할 때마다 api 요청이 필요한데, 의도한 것이 맞을까요?

jho2301 replied:
넵 로그인을 하지 않으면 서비스 메뉴를 통한 이동을 막도록 의도해봤습니다. 
그러려면 라우팅 단계에서 로그인 검증 로직이 있어야 한다고 생각해서 위 함수를 만들어봤습니다.

로그인이 필요한 페이지에서 사용하는 데이터 요청 api에 access토큰이 필요하긴 합니다. 
하지만 (기획상의 이유로)  첫 페이지 로딩 시, api가 곧바로 호출된다는 보장이 없고, api 호출이 필요 없는 정적인 데이터만 제공하는 페이지가 있을 수 있기에 라우터가드가 별도로 필요하다고 생각했습니다. 

반대로 login, signup 페이지는 로그인한 유저가 접근하면 안되는 페이지여서 로그인 상태라면 접근할 수 없도록 unauthorizedRoute라는 함수로 접근을 제한했습니다

jbee37142 reviewed (changes requested):
코멘트 달았어요~
코드가 전반적으로 깔끔해서 리뷰할게 별로 없네요! 👍 
PR body에 남겨주신 질문들이 추상적이고 어떤 고민에서 나온 질문들인지 잘 모르겠어서 역질문 드려봅니다.

---

코멘트 달았어요! :)

> 필요할 때마다 리팩토링을 통해 구조를 잡아 나갔습니다. 관련해서 팁이 있다면 조언 부탁드리고 싶습니다.

어떤 고민이 있는지 잘 모르겠네요! 필요할 때마다 리팩토링을 하는게 맞지 않을까요?

> addEventListener로 걸어준 이벤트 대상 DOM 객체가 삭제되거나 detach되서 더 이상 참조되지 않으면, 일일이 removeEventListener를 해주는게 좋을까요? 아래 글을 보았는데 애매해서 질문드립니다.

왜 해줘야 한다고 생각했는지 궁금하네요!

> SPA를 구현할 때, 다른 페이지의 HTML 코드를, 별도의 html 파일로 만드는 게 좋을까요? 아니면 js 파일에서 template 문자열로 가지고 있게 하는 것이 좋을까요?

각각의 장단점이 무엇이라고 생각하셔서 이 질문을 하게 된 것일까요?

> 위 코드는 html 파일을 fetch해 와서 문자열로 변환한 뒤, innerHTML로 렌더하는 예시 코드입니다. 이 형식이 갖는 이점과 template 문자열을 js로 관리해서 렌더하는 것과 어떤 차이가 있는지 궁금합니다.

이것도 무슨 말인지 잘 모르겠는데요, 일단 두 차이는 네트워크 비용이 든다는 부분일 것 같네요(?)

> html에 style 태그로 주입되는 것과, css 파일을 link 태그로 가져오는 것과 성능적으로나 기능적으로나 어떤 차이가 생기는지 궁금

브라우저가 Document를 어떻게 렌더링하는지 살펴보면 궁금증이 해결될 것 같은데요?

> 이전에는 html 파일과 css 파일을 따로 분리해서 개발하는 것이 당연하지만, 번들링을 해주니 굳이 css 파일을 분리해도 되는지 잘 모르겠습니다.

html 파일과 css 파일을 따로 분리해서 개발하는 것이 당연하다고 생각한 이유가 궁금하네요, 분리하는 것과 하지 않는 것 두 가지의 장단점이 뭐라고 생각하시고 왜 고민인지도 같이 말씀해주시면 더 좋은 질문이 될 것 같네요

jho2301 committed:
feat: 회원가입 시  email아이디 실시간 중복검사 및 form submit버튼 실시간 활성화

jho2301 commented:
### 기능 추가해보았습니다.

- 로그인 상태 유지 선택 기능
- 회원가입 시 email아이디 실시간 중복검사 기능
-  로그인, 회원가입 submit버튼 실시간 활성화 기능

위에 올려주신 질문은 좀 더 고민해보고 답글 달도록 하겠습니다. 피드백 감사드립니다!

jho2301 commented:
> 필요할 때마다 리팩토링을 통해 구조를 잡아 나갔습니다. 관련해서 팁이 있다면 조언 부탁드리고 싶습니다.

어떤 고민이 있는지 잘 모르겠네요! 필요할 때마다 리팩토링을 하는게 맞지 않을까요?

- 파노: 현재 어떤 고민이 있다기 보다는 리팩토링 관련해서 jbee님의 노하우가 있다면 이야기 들어보면 좋을 것 같아서 올린 부분이었습니다.

> addEventListener로 걸어준 이벤트 대상 DOM 객체가 삭제되거나 detach되서 더 이상 참조되지 않으면, 일일이 removeEventListener를 해주는게 좋을까요? 아래 글을 보았는데 애매해서 질문드립니다.

왜 해줘야 한다고 생각했는지 궁금하네요!

- 체프: 모든 브라우저의 GC가 동일하게 동작하지 않을 것이고, GC의 로직을 정확하게 알 수 없으니, 이벤트를 detach 해야하는 상황이 올 때, 코드로 확실하게 **명시**하는 것도 좋겠다는 생각을 했습니다.
- 파노: 성능상 문제가 올 수 있을 수 있을 것 같다. GC의 실행시점이 불분명한 것이 애매한 점. 현재는 문제가 되지 않는 부분이긴하다.

> SPA를 구현할 때, 다른 페이지의 HTML 코드를, 별도의 html 파일로 만드는 게 좋을까요? 아니면 js 파일에서 template 문자열로 가지고 있게 하는 것이 좋을까요? 위 코드는 html 파일을 fetch해 와서 문자열로 변환한 뒤, innerHTML로 렌더하는 예시 코드입니다. 이 형식이 갖는 이점과 template 문자열을 js로 관리해서 렌더하는 것과 어떤 차이가 있는지 궁금합니다.

각각의 장단점이 무엇이라고 생각하셔서 이 질문을 하게 된 것일까요?

- 파노: html템플릿태그를 html파일에서 관리하게 된다는점에서 인텔리센스나 코드포매팅이 용이하다는게 장점인 것  같습니다. 하지만 fetch를 해오고 다시 문자열로 파싱한다는점이 불필요한 비용일 수 있다는 생각이 들어 논의 끝에 js파일로 관리하기로 하였습니다. 관련해서 JBee님의 생각이 어떠신지 궁금했습니다.

> 이전에는 html 파일과 css 파일을 따로 분리해서 개발하는 것이 당연하지만, 번들링을 해주니 굳이 css 파일을 분리해도 되는지 잘 모르겠습니다.

html 파일과 css 파일을 따로 분리해서 개발하는 것이 당연하다고 생각한 이유가 궁금하네요, 분리하는 것과 하지 않는 것 두 가지의 장단점이 뭐라고 생각하시고 왜 고민인지도 같이 말씀해주시면 더 좋은 질문이 될 것 같네요

- 파노 : css를 파일 단위로 따로 분리하면 유지보수하기 편하다고 생각했습니다. 번들링을 해주니 네트워크 요청이 브라우저마다 다를 수 있지만 한 번에 최대 6개 파일을 가져온다고 알고있는데 파일을 하나로 번들링함으로써 네트워크 통신 최적화를 할 수 있는 점이 장점으로 생각됩니다.
- 체프 : 스타일시트 코드와 마크업 코드을 분리해서 개발할 수 있으며, 하나의 파일이 과도하게 커지는 것도 방지할 수 있습니다. 하지만 실제로 브라우저에서 렌더할 때는 css 파일과 html 파일을 따로따로 요청해서 가져와야 합니다.
네트워크 요청 시, 가져오는 파일의 개수와 요청은 최소화하는 것이 좋다고 생각합니다. 그래서 빌드할 때는 css 파일을 분리하지 않고 style 태그 안에 스타일시트 코드를 넣어서 한번에 빌드하는 것이 성능상으로 이점이 있다고 생각합니다. 마치 여러 개의 이미지 파일을 가져오기 위해 여러 번 요청하는 것이 아니라 이미지 스프라이트를 만들어 한번 요청으로 이미지를 가져오듯이 말입니다.
그래서 개발할 때는 파일을 분리하는 것의 이점이 더 크지만, 번들러가 있으면 굳이 번들러에서 빌드할 때도 css 파일을 분리해서 빌드할 필요가 있을까? 하는 의문이 생겼습니다.

jho2301 committed:
chore: babelrc 파일 삭제

jho2301 committed:
feat: authenticatedRoute strict 모드 추가

- strict 모드: api호출을 통해 로그인중인지 확인
- sloppy 모드: 프로세스 내 상태값에 access token이 있는지 확인

Co-authored-by: Shim MunSeong <puterism.k@gmail.com>

jho2301 commented:
관련 커밋: 28b0ce1

라우트 가드가 매번 api요청을 하는게 비효율적이라는 의견에 고민을 계속 해봤습니다.

그래서 authenticatedRoute는 액세스토큰 상태값이 있는지만 확인하는 걸 기본으로하고
특별히 라우트가드가 로그인여부를 서버에서 검증해야하면 옵션을통해 api를 호출하는 검증으로 사용하도록했습니다.

정상적인 상황에서는 상태값 체크하는 과정에서 걸러질 수 있고, 비정상적인 상황에서는 액세스토큰이 저장소에 남아 있다하더라도 액세스 토큰이 필요한 api호출 시 걸러질 것으로 기대합니다.

1차로 액세스 토큰이 있는지 확인 후, 없다면 로그인화면으로 리다이렉션
2차로 api 호출 -> 에러 시 로그인화면으로 리다이렉션 처리를 하였습니다

jbee37142 commented:
> 파노: 현재 어떤 고민이 있다기 보다는 리팩토링 관련해서 jbee님의 노하우가 있다면 이야기 들어보면 좋을 것 같아서 올린 부분이었습니다!

리팩토링에 팁 같은건 없어요 ㅎㅎ

> 메모리 해제

GC에 맡기지 않고 매번 수동으로 해제해주는 것은 트레이드 오프가 있다고 생각해요. 힙 메모리 분석을 통해서 정말로 메모리 누수가 발생하는 경우를 검증하고 작업에 들어가도 늦지 않을 것 같다는 것이 제 생각입니다.


> 파노: html템플릿태그를 html파일에서 관리하게 된다는점에서 인텔리센스나 코드포매팅이 용이하다는게 장점인 것 같습니다. 하지만 fetch를 해오고 다시 문자열로 파싱한다는점이 불필요한 비용일 수 있다는 생각이 들어 논의 끝에 js파일로 관리하기로 하였습니다. 관련해서 JBee님의 생각이 어떠신지 궁금했습니다.

HTML 을 fetch 해온다는 것이 무슨 말인지 잘 모르겠지만 정적으로 관리하는 것이 좋은 것 같아요!

> 이전에는 html 파일과 css 파일을 따로 분리해서 개발하는 것이 당연하지만, 번들링을 해주니 굳이 css 파일을 분리해도 되는지 잘 모르겠습니다.

말씀하신대로 분리해서 개발하고 번들링해주면 되지 않을까요...? 답변을 이미 알고 계신 것 같은데 ㅎㅎ

추가로, 체프님의 네트워크 비용 최적화 관련된 내용은 반은 맞고 반은 틀린 내용이긴 하지만 정말 훌륭하네요!

jbee37142 reviewed (approved):
질문에 대한 코멘트 달았어요!
